### ISSUE TITLE
Mouse Cursor / Pointer / Sprite Keeps Disappearing 

### ISSUE THREAD

**Issue by PunchCardReader** [XQuartz/XQuartz#208]
Dear Madam or Sir, 

I regret to report a bug in XQuartz. I hesitate to call it a bug, since it could be an interaction with macOS. 
In this bug report I will lay out precise conditions for triggering the problem. 

I’m a lifelong software developer. I have used X11 terminals in some form for 30 years. I have some experience building GUI apps in Motif. 
So I’m surprised that it took me so long (9 months) to identify and refine the trigger for this problem. 

**Symptom:** Mouse Cursor / Pointer / Sprite Disappears when switching context away from XQuartz xterm windows. 
User begins randomly flailing, trying everything to get the mouse pointer to reappear. 

**Platform:** Mac OS X 10.11.6 (El Capitan), Intel processor 

**Versions Affected:** 
• whatever version I was using in December 
• XQuartz 2.8.0 (install.log calls it 2.8.10), installed 9 Mar 2021 
• XQuartz 2.8.1 (install.log calls it 2.8.19), installed 10 Sep 2021 

**Workaround:** If the problem gets triggered, this solution always works for me: Hold down the Command button. Then hit the Tab key and release it, and then hit the ‘`’ (back-quote) key and release it (the one near the upper left corner of the keyboard). Then release the Command button. This sequence is intended to pop up the macOS App Switcher, briefly switch to the next app in the sequence, then switch back to the app I was in when I noticed the pointer was missing. 

**Interaction with Rollover:** 

**Very Important Clue:** **The effect is only triggered in situations involving 2-key rollover.** 
In other words, it makes a difference which keyboard button you release first. 
“Rollover” is a computer design term from many decades ago; for background, please visit https://en.wikipedia.org/wiki/Rollover_(key) . 

As an example: If I press the Command key, then the Tab key, and then I release the Command key *first*, and release the Tab key *last*, that’s a **rollover** situation, 
as though I’m literally *rolling* my hand over the two keys as I hit them. 
A **non-rollover** situation means releasing the Tab key first, “in LIFO order” or “out of sequence”, one might say; tapping the Tab key while the Command key is down. 

Imagine we attach alligator clips to the metal contacts on the backs of the keyboard switches, and graph the voltage levels across the switches in an oscilloscope or strip-chart recorder while typing. 
A **rollover** would look like this:

![rollover](https://user-images.githubusercontent.com/90868761/133687553-721e74cb-b46c-4dff-a28e-616e846444de.png)

and a **non-rollover** would look like this:

![non-rollover](https://user-images.githubusercontent.com/90868761/133687603-d81837d9-9eac-4891-985d-595be32f908f.png)

Mac GUI programs treat these situations identically: In most desktop apps, it makes no difference which key you release first. Thus we say Mac GUI programs are not sensitive to keyboard rollover. Apple desktop machines handle both styles of typing gracefully. Likewise for all the XWindows workstations I’ve used. 
As a result, computer users have not had to learn to control their rollovers. (Nor should they have to — let me be clear.) 
When I’m typing rapidly, I randomly generate *both* rollovers and non-rollovers; 
this could explain why the bug seems to appear at random. When I controlled my rollovers, I gained the ability to scientifically test this bug. 

**tl;dr: XQuartz xterms have an interaction with the macOS App Switcher which causes Command-Tab rollovers to make the pointer disappear. Also, switching to another Space makes the pointer disappear.** 

I confirmed the rollover trigger using an old Apple Design keyboard from the 1990s, as well as an iMac keyboard from the 2000s, both desktop keyboards. 
I have not tried this on laptops nor touch screens. 

**Trigger #1: How to Reproduce the Effect:** 

1. You will need XQuartz running with a few of its own xterm windows (not Apple’s Terminal.app, but XQuartz’s Command-N Terminals). 

2. Place your mouse pointer over the topmost (active) XQuartz xterm window, i.e., whichever xterm has the focus. (This step turns out to be crucial.) 

3.a. Carefully type Command-Tab using the two-key rollover technique. 

3.b. For comparison, type Command-Tab using the non-rollover technique. 

The rollover technique produces the disappearing pointer effect with high probability. 
The non-rollover technique leaves the pointer visible with near certainty. 
Important: _Because probability is involved, please do not delegate this bug report to a junior staff member who does not have a degree in mathematics or computer science._ Non-mathematical persons tend to test something *once* and declare it solved. 
Also, when patching this bug, do multiple tests, and don’t be discouraged by an occasional isolated failure. 

**Trigger #2: How to More Easily Reproduce the Effect:** 

The xterm is also involved in a feature interaction with macOS’s “Spaces” feature. 

Starting in XQuartz, place the mouse pointer over the frontmost xterm window. 
Use Control-RightArrow or Control-LeftArrow to switch Spaces. Or, if you have assigned accelerator keys to your Spaces, you may use those. 
The mouse pointer disappears. 

This method works a little differently: The macOS App Switcher does not pop up, and the disappearance of the pointer is *unconditional*, i.e., the pointer disappears whether or not you use the rollover technique. 
Because this trigger is so easy to reproduce, you may wish to use it when testing. 

**How to Get the Pointer to Reappear:** 

Hit Command-Tab, then Command-` (aka Command-Backquote, Command-AccentGrave). 

**What Does *Not* Trigger the Effect:** 

• In XQuartz, place the mouse pointer over a non-front window, or over the background (on a Mac, this part of the screen will be either the Desktop or the visible windows of a non-front app). In this case, switching to another app does not trigger the pointer to disappear. 

• HOWEVER: Recall where I used the word “crucial” above. If you open XQuartz Preferences, click the “Windows” tab, and check the box labeled “Focus Follows Mouse”, now (in effect) *every* xterm window is the frontmost window, and satisfies the conditions for triggering the effect. 

![Screen Shot 2021-09-16 at 3 25 11 PM](https://user-images.githubusercontent.com/90868761/133687891-65ae920c-ef53-4840-ac59-2a99a831d549.png)

**Possible Software Solution:** 

Correcting this problem could be tricky. 
If you practice your rollovers and your non-rollovers in a utility like “xev” 
(you’ve seen xev, the old X11 printer of XEvents), 
you can see the exact sequence of KeyPress and KeyRelease events. 
Try this in xev, using Command-C (the copy command. That should be safe enough, safer than Command-X and Command-V). 

Problem is, Command-Tab doesn’t register in XQuartz apps. 
As you will see when you try Command-Tab in xev, the macOS App Switcher “steals” the Tab key for its own use, 
which keeps the Tab KeyPress event from reaching xev. 
I assume xterm never gets the Tab key event either. 
This would seem to make it impossible for XQuartz and xterm to detect a Command-Tab, say, for purposes of un-hiding the pointer during context switches. 

**Best Guess:** Solutions might require that XQuartz or xterm watch for a FocusOut event while the Command key is down. Waiting for a Tab key may not work. 
Rollover detection shouldn’t be necessary; however … why the rollover is a necessary part of Trigger #1 is an unsolved mystery. 
You may have better luck patching Trigger #2. 

• Am I blaming XQuartz? I will simply point out that if I click Command-H to hide XQuartz, and spend a few minutes just switching back and forth between Firefox and Apple’s apps, the problem does not occur. 

**Notes & References:** 

1. According to the good people at Apple Toolbox, the disappearing pointer problem is known as the “Where’s My Mouse?” Syndrome. They describe very much what I’m seeing, and the method for getting the pointer to reappear, but without naming a trigger. See https://appletoolbox.com/mouse-cursor-pointer-disappears-invisible-missing-fix/ . 

2. Note to self: In my notes I named this bug “Bug 121720-P”, for the date I first investigated it, 17 Dec 2020.

**Comment by jeremyhu** [comment]
Also see https://github.com/XQuartz/XQuartz/issues/100

**Comment by jeremyhu** [comment]
xterm hides the X11 cursor when you start typing, so that is why it is disappearing (correctly).  The problem here is that it isn't being *unhidden* when leaving the X11 context.  As such, it looks like this actually is a dupe of #100 (which was fixed in 2.8.0), but you state here that you are still seeing this with 2.8.0 and 2.8.1.  I tried both cases you outlined above (on macOS 12.5) multiple times and the pointer always appeared.

Quite puzzling.

It's possible that there was also a macOS bug as well (in addition to #100) which has since been fixed.  I'm curious if you still see the issue with current versions of macOS (eg: either 12.5 or 13.0 betas).
